Make trait properties and methods first class node citizens

herndlm opened the pull request:
This is basically the refactoring I had in mind in https://github.com/phpstan/phpstan-src/pull/1362

The current state is already a bit weird IMO, in general trait props and methods were ignored when gathering class nodes. Then recently I added trait props to a dedicated array, just to use them also via uninitialized props checks. I did not add them to the normal props because that would lead to annoying new errors via `UnusedPrivatePropertyRule`.
Then just yesterday I added trait property promotion usage to the normal property usage, but ignoring other potential trait prop usages elsewhere. And trait methods are still completely ignored.

So what does this do different now?
Trait props and methods are gathered together with their class counterparts. And rules just use that. Additionally the information if the prop or method came from a trait is also remembered and only that is used to skip trait props / methods in those 2 `UnusedPrivate*` rules, that were annoying.

I think in the long-run this makes more sense and avoids more weird trait problems. But maybe the details could be modified a bit. WDYT?

herndlm committed:
Make trait properties and methods first class node citizens

herndlm commented:
Aah I guess I could keep BC here still if really needed ;)
I added that method in https://github.com/phpstan/phpstan-src/pull/1340 which was released in 1.7.0 a week ago. Do you think we need it?

Oh and sorry for the title rename. I messed it up..

ondrejmirtes commented on src/Node/ClassStatementsGatherer.php:
```diff
@@ -66,11 +73,11 @@ public function getProperties(): array
+	 * @return array<int, int>
 	 */
-	public function getTraitProperties(): array
+	public function getIsTraitProperty(): array
```
I don't understand this API at all :) What's `array<int, int>`?

It'd be much nicer to add this information `ClassPropertyNode` - a new `isDeclaredInTrait` method.

herndlm replied:
agreed, it's a bit weird to use / understand. this is basically what `isOptionalProperty` is in `ConstantArrayType`, so it works with the numeric index. Such a method would be way nicer though, agreed. Will take a look at it in the evening.

ondrejmirtes replied:
Yeah, I don't like that either but I don't know how else it should work there while still being fast. We can afford to do something nicer in this case.

ondrejmirtes commented on src/Node/ClassStatementsGatherer.php:
```diff
@@ -81,6 +88,14 @@ public function getMethods(): array
+	/**
+	 * @return array<int, int>
+	 */
+	public function getIsTraitMethod(): array
```
We can change the return type of `getMethods` to our own object and also add an information about a trait there.

herndlm replied:
with our own object, you mean some new object basically or something different?

ondrejmirtes replied:
A new class, ClassStatemenetsGatherer contains many of those. They don't even have to be a `Node`, for example https://github.com/phpstan/phpstan-src/blob/1.7.x/src/Node/Method/MethodCall.php isn't.

herndlm commented:
I added a new virtual node `ClassMethodNode` and added `isDeclaredInTrait` to it and to `ClassPropertyNode`. In general I like that approach, but there are a couple of open questions for those two classes:
- Currently I only need the name, private visibility yes/no and the trait info. But I already started duplicating other `ClassMethod` methods like `isPublic`, `isStatic`, .. how far should we go here? it's not complete yet either, e.g. `getParams`, `getReturnType`, .. are all still missing. I don't like the intermediate state
- The duplicated methods are, well, duplicated. Would it make more sense to either store `$originalNode` or use it in the constructor to get the needed info? Or should as much as possible be injected via constructor?
- I assume `ClassPropertyNode` grew a bit over time, e.g. it didn't have the `$originalNode` in it's initial state, whatever we do here, should we do this to `ClassPropertyNode` too then? Maybe for consistency reasons a bit of cleanup afterwards would be good :) but currently I don't fully understand where to go with those 2 classes exactly.

I guess my first intuition would be to kind of decorate a `ClassMethod`, meaning to hold one internally and support the same methods by passing everything through. Or just expose that internal node completely. But additionally add my custom method. Is this problematic memory-wise I assume?

UPDATE: I was so focused on `ClassPropertyNode` that I completely overlooked your other example and the fact that they _don't_ have to be a real node.. Currently cleaning this up heavily :) And I guess ClassPropertyNode can't be cleaned-up easily, looks like it's in BC-lockdown mode

herndlm commented on src/Node/ClassMethod.php:
```diff
@@ -0,0 +1,26 @@
+	)
+	{
+	}
+
```
```suggestion
```

herndlm replied:
OH NO, BUT WHAT ABOUT MY NEWLINE ;)

ondrejmirtes replied:
I didn't understand that :) There should be a PHPCS sniff catching that (https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesmethodspacing-), feel free to enable that and send a new PR :)

herndlm replied:
this is really bothering me, so yes, I'll do that :)

ondrejmirtes commented:
Nice, thank you!
